feature/AddListBottomSheet

Mostafa-alsaygh opened the pull request:
# What does this PR do?

This PR introduces functionality for adding, as well as viewing items within a list.

Mostafa-alsaygh committed:
feat(list): add edit list bottom sheet

This commit introduces the `EditListBottomSheet` composable, which allows users to edit the name of a list.

The bottom sheet includes:
- A title "Edit list".
- A text field for the list name with a leading icon.
- A "Save" button to apply changes.
- A close button.

String resources for "Edit list", "Save", and "List title" have been added in both English and Arabic.
A new drawable icon `editlisticon.xml` has been added for the list name text field.

The following files related to the previous `ViewListItems` feature have been removed as this functionality is being replaced or refactored:
- `ItemsType.kt`
- `MediaUi.kt`
- `ViewListItemsContract.kt`
- `ViewListItemsEffect.kt`
- `ViewListItemsScreen.kt`
- `ViewListItemsUiState.kt`
- `ViewListItemsViewModel.kt`

New files for the list feature include:
- `EditListContract.kt`
- `ListEffect.kt`
- `ListUiState.kt`
- `ListViewModel.kt`

Mostafa-alsaygh committed:
Merge branch 'refs/heads/develop' into feature/edit-list-bottom-sheet

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/list/viewlistitems/ViewListItemsScreen.kt
#	presentation/src/main/res/values-ar/strings.xml
#	presentation/src/main/res/values/strings.xml

Mostafa-alsaygh committed:
Merge branch 'refs/heads/develop' into feature/edit-list-bottom-sheet

Mostafa-alsaygh committed:
feat(list): Add/Edit list bottom sheet and view list items screen

yusufnasserdev commented on presentation/src/main/res/drawable/editlisticon.xml:
File name doesn't follow the resources naming convention, it should be something like `ic_edit_list`.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/viewlistitems/ViewListItemsScreen.kt:
```diff
@@ -63,11 +64,9 @@ fun ViewListItemsScreen(
+        isError = state.error is ErrorState.NoInternet,
         pagingFlow = listItems,
-        handlePagingLoadingAutomatically = false
+        handlePagingLoadingAutomatically = true
```
These changes are outside the scope of this PR, what are you doing here?

Mostafa-alsaygh replied:
i did not change on these  lines there is a problem in github when i merge from develop to my branch

yusufnasserdev replied:
They are changes in your branch, how a GitHub Pull Request work is they show you the difference between your branch's latest commit (head) and the base branch latest.

Kindly check this further.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListUiState.kt:
```diff
@@ -10,4 +12,19 @@ data class ListUiState(
+    val listId: String? = null,
+    val sheetMode: ListSheetMode = ListSheetMode.ADD,
+    val mediaType: MediaType = MediaType.Movie,
+    val isSheetLoading: Boolean = false,
```
The sheet state should be extracted into its own separate class to allow for encapsulation.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListScreen.kt:
```diff
@@ -61,6 +63,8 @@ fun ListScreen(
     effect?.Listen { currentEffect ->
         when (currentEffect) {
             is ListEffect.NavigateToDetails -> onNavigateToDetails(currentEffect.id)
+            ListEffect.ShowAddListSheet -> TODO()
```
Missing `is`

ahmeddsamii replied:
no need for is here

when if data class we need is but if it is object or data object we don't need is @yusufnasserdev but the real problem here is TODO().... it will cause a crash

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListScreen.kt:
```diff
@@ -322,6 +326,17 @@ private fun Preview() {
+                override fun onMediaTypeChanged(mediaType: MediaType) {}
+                override fun showAddListSheet(mediaType: MediaType) {}
+                override fun showEditListSheet(
+                    listId: String,
```
Id as a String?

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListEffect.kt:
```diff
@@ -1,5 +1,11 @@
 sealed interface ListEffect {
     data class NavigateToDetails(val id: Int) : ListEffect
+    object ShowAddListSheet : ListEffect
+    data class ShowEditListSheet(val listId: String, val name: String, val mediaType: MediaType) :
```
Id as a String?

Mostafa-alsaygh replied:
the  id for the list i make it string, what is yours suggestion

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListContract.kt:
```diff
@@ -1,8 +1,19 @@
+    fun onListNameChanged(listName: TextFieldValue)
+    fun onMediaTypeChanged(mediaType: MediaType)
+    fun showAddListSheet(mediaType: MediaType = MediaType.Movie)
+    fun showEditListSheet(listId: String, currentName: String, mediaType: MediaType)
```
Check this Id here too.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+            override fun onListNameChanged(listName: TextFieldValue) {}
+            override fun onMediaTypeChanged(mediaType: MediaType) {}
+            override fun showAddListSheet(mediaType: MediaType) {}
+            override fun showEditListSheet(
```
Create a default contract object and pass it to the reviews.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+            override fun onListNameChanged(listName: TextFieldValue) {}
+            override fun onMediaTypeChanged(mediaType: MediaType) {}
+            override fun showAddListSheet(mediaType: MediaType) {}
+            override fun showEditListSheet(
```
Create a default contract object and pass it to the reviews.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+                if (!listUiState.isLoading) {
+                    editInteractions.onEditListSheetDismiss()
+                }
+            },
```
You don't hide the sheet on dismiss request?

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .wrapContentHeight()
+                    .heightIn(max = LocalWindowInfo.current.containerSize.height.dp * 0.35f)
+                    .padding(bottom = 24.dp)
```
Why is this needed?

Mostafa-alsaygh replied:
it is padding like a figma same as the filter bottom sheet that deleted

yusufnasserdev replied:
The question was about the heightIn, you don't need to force a height on the bottom sheet, it determines its height from its components.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    val isButtonEnabled = when (listUiState.sheetMode) {
+        ListSheetMode.ADD -> !listUiState.isLoading &&
+                listUiState.listName.text.trim().isNotEmpty()
+
+        ListSheetMode.EDIT -> !listUiState.isLoading &&
+                listUiState.listName.text.trim().isNotEmpty() &&
+                listUiState.listName.text.trim() != listUiState.originalListName
```
Extract each condition to a function that describe its value.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+            containerColor = NovixTheme.colors.surface,
+            state = sheetState,
+        ) {
+            Box(
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .wrapContentHeight()
+                    .heightIn(max = LocalWindowInfo.current.containerSize.height.dp * 0.35f)
+                    .padding(bottom = 24.dp)
```
You can pass this modifier to the `ModalBottomSheet` component itself, you don't need to created a new layout.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    Column(
+        modifier = modifier.padding(horizontal = 16.dp)
```
Add `verticalArrangement = Arrangement.spacedBy(24.dp)` instead of padding it to each component.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+        Row(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(top = 8.dp, bottom = 24.dp),
```
There is no top padding, bottom padding should be spacing handled by the column.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+                        enabled = !listUiState.isLoading,
+                        onClick = onCloseClicked
+                    )
+                    .padding(6.dp),
```
There is no padding 6 dp anywhere in the design of this button.

Mostafa-alsaygh replied:
the icon will be not right without it

london-bot[bot] committed:
Branch was auto-updated.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+                    .padding(6.dp),
+                painter = painterResource(com.london.designsystem.R.drawable.cancel),
+                contentDescription = "Close",
+                tint = if (listUiState.isLoading)
+                    NovixTheme.colors.title.copy(alpha = 0.5f)
+                else
+                    NovixTheme.colors.title
```
Where in the design is this?

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+            )
+        }
+
+        Text(
+            text = stringResource(R.string.list_title),
+            style = NovixTheme.typography.title.small,
+            color = NovixTheme.colors.title,
+            modifier = Modifier.padding(bottom = 8.dp)
+        )
+
+        OutlinedTextField(
+            value = listUiState.listName,
+            interactionSource = interactionSourceUserName,
+            onValueChange = editInteractions::onListNameChanged,
+            leadingIcon = painterResource(R.drawable.editlisticon),
+            isError = listUiState.errorMessage != null,
+            enabled = !listUiState.isLoading
+        )
```
These should be in one component, probably a column.

Mostafa-alsaygh replied:
there is a main column why for the nested

yusufnasserdev replied:
Because together, they represent a single piece of ui, list title and its text field.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddEditListBottomSheet.kt:
```diff
@@ -0,0 +1,261 @@
+            onClick = onActionClicked,
+            enabled = isButtonEnabled,
+            modifier = Modifier
+                .padding(top = 24.dp)
```
No need for this padding.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
refactor: Improve list feature UI and state management

This commit refactors the list feature with the following changes:

- Introduces default contract implementations for `ListContract` and `ViewListItemsContract` to simplify preview and testing setups.
- Renames `onEditListSheetDismiss` to `onEditAddListSheetDismiss` in `ListContract` and `ListViewModel` for clarity.
- Modifies `ShowEditListSheet` effect in `ListEffect` to accept an `Int` for `listId` instead of a `String`.
- Updates `AddEditListBottomSheet` to use the new `editAddInteractions` and `editAddSheetState` for better state management.
- Refactors `ListUiState` by moving sheet-specific properties into a new `EditAddSheetState` data class.
- Renames the drawable `editlisticon.xml` to `ic_edit_list.xml` for consistency.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Refactor: Extract functions for title, button text, and button enabled state in AddEditListBottomSheet

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/feature/edit-list-bottom-sheet' into feature/edit-list-bottom-sheet

Mostafa-alsaygh committed:
Refactor: Update OutlinedTextField to use label for list title

Mostafa-alsaygh committed:
refactor: Remove edit list functionality and bottom sheet from saved list screen

The following changes were made:
- Removed `ShowEditListSheet` from `ListEffect`.
- Removed `onEditAddListSheetDismiss`, `onSaveEdit`, and `showEditListSheet` from `ListContract` and its default implementation.
- Renamed `AddEditListBottomSheet.kt` to `AddListBottomSheet.kt`.
- Updated `AddListBottomSheet` to only support adding new lists.
- Removed `isEditBottomSheetVisible` from `ViewListItemsUiState`.
- Removed `showEditListSheet` and related logic from `ListViewModel`.
- Removed `editAddSheetState` and `ListSheetMode` from `ListUiState`.
- Renamed `ic_edit_list.xml` to `ic_save_list.xml`.

Mostafa-alsaygh committed:
Merge branch 'develop' into feature/edit-list-bottom-sheet

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListViewModel.kt:
```diff
@@ -16,21 +18,42 @@ class ListViewModel @Inject constructor() :
         emitEffect(ListEffect.NavigateToDetails(id))
     }
 
     override fun onFabClick() {
-       // TODO("Not yet implemented")
+        /*TODO*/
```
What is this change?

Mostafa-alsaygh replied:
to not make crash i think

Mostafa-alsaygh committed:
fix:revert

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/feature/edit-list-bottom-sheet' into feature/edit-list-bottom-sheet

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/savedlist/ListUiState.kt:
```diff
@@ -10,4 +12,16 @@ data class ListUiState(
     val items: Flow<PagingData<ListItemUi>> = flow {},
-)
\ No newline at end of file
+    val isSheetVisible: Boolean = false,
```
Should be included within the sheet state.

Mostafa-alsaygh replied:
the visibly should be in the Ui state because it is a part of screen

yusufnasserdev replied:
It's a part of the state sheet which is accessible from within the ListUiState

Mostafa-alsaygh replied:
when you are in screen u dont need to know if this screen is turn of off in its ui state ... same thing in this state, what is the benefit of it to know

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddListBottomSheet.kt:
```diff
@@ -0,0 +1,191 @@
+
+@Composable
+@Preview
+fun PreviewAddMode() {
```
Change to `Preview` as per the team's conventions.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddListBottomSheet.kt:
```diff
@@ -0,0 +1,191 @@
+    modifier: Modifier = Modifier,
+    addListInteractions: ListContract = hiltViewModel<ListViewModel>(),
+    sheetState: SheetState = rememberModalBottomSheetState(),
+    listUiState: ListUiState,
```
The `listUiState` shouldn't be exposed to the sheet since the sheet has its own state within it.

Mostafa-alsaygh replied:
it should to control the visibly, the visibly should be in the Ui state because it is a part of screen

yusufnasserdev replied:
You have created a State for the sheet, which is `data class AddSheetState`, you should pass it here not the entire screen state.

yusufnasserdev reviewed (commented):
Please check earlier comments, kindly leave comments unresolved and I'll resolve them while re-reviewing.

Mostafa-alsaygh committed:
refactor: remove unnecessary height constraint and rename preview function in AddListBottomSheet

Mostafa-alsaygh committed:
Merge branch 'develop' into feature/edit-list-bottom-sheet

# Conflicts:
#	presentation/src/main/res/values-ar/strings.xml

Mostafa-alsaygh committed:
Remove unused "save" string resource

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Refactor: Decouple AddListBottomSheet visibility from ListUiState

The visibility of the `AddListBottomSheet` is no longer managed by `ListUiState`. Instead, it is now controlled directly by the `AddSheetState`.

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/feature/edit-list-bottom-sheet' into feature/edit-list-bottom-sheet

Mostafa-alsaygh committed:
Merge branch 'develop' into feature/edit-list-bottom-sheet

# Conflicts:
#	presentation/src/main/res/values-ar/strings.xml

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddListBottomSheet.kt:
```diff
@@ -0,0 +1,187 @@
+                    .padding(bottom = 24.dp)
+            ) {
+                AddListBottomSheetContent(
+                    modifier = modifier,
```
You can't pass the same modifier to two components.

yusufnasserdev commented on presentation/src/main/java/com/london/presentation/feature/list/bottomsheets/AddListBottomSheet.kt:
```diff
@@ -0,0 +1,187 @@
+            state = sheetState,
+        ) {
+            Box(
+                modifier = modifier
+                    .fillMaxWidth()
+                    .wrapContentHeight()
+                    .padding(bottom = 24.dp)
```
You can pass this modifier to the ModalBottomSheet and skip the box (extra layout).

Mostafa-alsaygh committed:
Refactor: Remove redundant Box in AddListBottomSheet

yusufnasserdev reviewed (approved):
Awesome work, keep it up brother 🔥
